### PR TITLE
Inconsistency for dimension between R and backend. fix #544

### LINKED ISSUE #544: R regression example does not work
In a fresh R session:
```{R}
library(mxnet)
data(BostonHousing, package="mlbench")
train.ind = seq(1, 506, 3)
train.x = data.matrix(BostonHousing[train.ind, -14])
train.y = BostonHousing[train.ind, 14]
test.x = data.matrix(BostonHousing[-train.ind, -14])
test.y = BostonHousing[-train.ind, 14]
data <- mx.symbol.Variable("data")
fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)
lro <- mx.symbol.LinearRegressionOutput(fc1)
mx.set.seed(0)
model <- mx.model.FeedForward.create(
  lro, X=train.x, y=train.y,
  ctx=mx.cpu(), num.round=50, array.batch.size=20,
  learning.rate=2e-6, momentum=0.9, eval.metric=mx.metric.rmse)
```

Yields:
```{R}
Auto detect layout of input matrix, use rowmajor..
Start training with 1 devices
[15:07:01] ./dmlc-core/include/dmlc/logging.h:208: [15:07:01] src/ndarray/ndarray.cc:159: Check failed: from.shape() == to->shape() operands shape mismatch
 Hide Traceback
 
 Rerun with Debug
 Error: [15:07:01] src/ndarray/ndarray.cc:159: Check failed: from.shape() == to->shape() operands shape mismatch 
5 stop(structure(list(message = "[15:07:01] src/ndarray/ndarray.cc:159: Check failed: from.shape() == to->shape() operands shape mismatch", 
    call = NULL, cppstack = NULL), .Names = c("message", "call", 
"cppstack"), class = c("Rcpp::exception", "C++Error", "error", 
"condition"))) 
4 exec$update.arg.arrays(arg.arrays, match.name, skip.null) 
3 mx.exec.update.arg.arrays(train.execs[[i]], s, match.name = TRUE) 
2 mx.model.train(symbol, ctx, input.shape, params$arg.params, params$aux.params, 
    1, num.round, optimizer = optimizer, train.data = X, eval.data = eval.data, 
    metric = eval.metric, epoch.end.callback = epoch.end.callback, 
    batch.end.callback = batch.end.callback, kvstore = kvstore) 
1 mx.model.FeedForward.create(lro, X = train.x, y = train.y, ctx = mx.cpu(), 
    num.round = 50, array.batch.size = 20, learning.rate = 2e-06, 
    momentum = 0.9, eval.metric = mx.metric.rmse) 
```

Their is either a bug in the code or with my installation.  The sonar classification example and MNIST examples work fine.  I followed the install guide here:
https://medium.com/@lelayf/installing-mxnet-for-r-on-yosemite-291470055980

```{R}
> sessionInfo()
R version 3.2.2 (2015-08-14)
Platform: x86_64-apple-darwin14.5.0 (64-bit)
Running under: OS X 10.10.5 (Yosemite)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] mxnet_0.5

loaded via a namespace (and not attached):
 [1] htmlwidgets_0.5  DiagrammeR_0.8.1 magrittr_1.5     visNetwork_0.1.2 htmltools_0.2.6  tools_3.2.2     
 [7] rstudioapi_0.3.1 Rcpp_0.12.1      codetools_0.2-14 stringi_1.0-1    data.table_1.9.6 jsonlite_0.9.17 
[13] stringr_1.0.0    digest_0.6.8     chron_2.3-47    
```

### PR DESCRIPTION
**Pull request by thirdwing**

### CODE REVIEW

**Comment by tqchen** [comment]
@thirdwing You might want to revert this. The problem was due to commit by @piiswrong  in infer shape

**Comment by tqchen** [comment]
This will break the softmax classification case, as it seems

**Comment by thirdwing** [comment]
Yes, it fixed the RMSE and broke softmax. You can revert it or give me a little time to fix both.
